doc/man3/OSSL_PARAM_TYPE.pod

```diff
@@ -0,0 +1,320 @@
+
+OSSL_PARAM_get_TYPE() retrieves a value of type B<TYPE> from the parameter B<p>.
+The value is copied to the address B<val>.
+Type coercion takes place as discussed in the NOTES section.
+
+OSSL_PARAM_set_TYPE() stores a value B<val> of type B<TYPE> into the paramter
+B<p>.
+Type coercion takes place as discussed in the NOTES section.
+
+OSSL_PARAM_get_BN() retrieves a BIGNUM from the parameter pointed to by B<p>.
+The BIGNUM referenced by B<val> is updated and is allocated if B<*val> is
+B<NULL>.
+
+OSSL_PARAM_set_BN() stores the BIGNUM B<val> into the paramater B<p>.
+
+OSSL_PARAM_get_utf8_string() retrieves a UTF8 string from the parameter
+pointed to by B<p>.
+The string is either stored into B<*val> with a length limit of B<max_len> or,
+in the case when B<*val> is B<NULL>, memory is allocated for the string and
+B<max_len> is ignored.
+If memory is allocated by this function, it must be freed by the caller.
+
+OSSL_PARAM_set_utf8_string() sets a UTF8 string from the parameter pointed to
+by B<p> to the value referenced by B<val>.
+
+OSSL_PARAM_get_octet_string() retrieves an OCTET string from the parameter
+pointed to by B<p>.
+The OCTETs are either stored into B<*val> with a length limit of B<max_len> or,
+in the case when B<*val> is B<NULL>, memory is allocated and
+B<max_len> is ignored.
+If memory is allocated by this function, it must be freed by the caller.
+
+OSSL_PARAM_set_octet_string() sets an OCTET string from the parameter
+pointed to by B<p> to the value referenced by B<val>.
+
+OSSL_PARAM_get_utf8_ptr() retrieves the UTF8 string pointer from the parameter
+referenced by B<p> and stores it in B<*val>.
+
+OSSL_PARAM_set_utf8_ptr() sets the UTF8 string pointer in the parameter
+referenced by B<p> to the values B<val>.
+
+OSSL_PARAM_get_octet_ptr() retrieves the OCTET string pointer from the parameter
+referenced by B<p> and stores it in B<*val>.
+The length of the OCTET string is stored in B<*used_len>.
+
+OSSL_PARAM_set_octet_ptr() sets the OCTET string pointer in the parameter
+referenced by B<p> to the values B<val>.
+The length of the OCTET string is provided by B<used_len>.
+
+=head1 RETURN VALUES
+
+OSSL_PARAM_construct_TYPE(), OSSL_PARAM_construct_BN(),
+OSSL_PARAM_construct_utf8_string(), OSSL_PARAM_construct_octet_string(),
+OSSL_PARAM_construct_utf8_ptr() and OSSL_PARAM_construct_octet_ptr()
+return a populated OSSL_PARAM structure.
+
+OSSL_PARAM_locate() returns a pointer to the matching OSSL_PARAM object.
+It returns B<NULL> on error or when no object matching B<key> exists in
+the B<array>.
+
+All other functions return B<1> on success and B<0> on failure.
+
+=head1 NOTES
+
+Automatic type coercion takes places between the various sizes of types using
+the C conversion rules.
+Integral types  will be widened, sign extended and narrowed as required.
+Unsigned and signed integer types will be automatically converted.
+During widening, signed values will be sign extended and unsigned values
```

Comment:
Are we still going to do this?

Reply:
This is what the code does now

Reply:
I can't readily follow the code, so can you answer this question?  What happens if I do a get_int call on a parameter which was set to be too big?

Reply:
Truncation.  Here is a line that should answer that question precisely:

``` C
        *val = (int)*(const int64_t *)p->buffer;
```

|val| is defined like this, and represent goal the object attribute: `int *val`

In this case, the param data type indicates a 64-bit value.  So what happens is that the 64-bit value is taken from the buffer, then cast to an int, before it ends up in whatever |val| points at.

Reply:
Ah, sorry, I was wrong, truncation is *not* happening...  That line is wrapped like this:

``` C
        if (sizeof(int) >= sizeof(int64_t)) {
            *val = (int)*(const int64_t *)p->buffer;
            return 1;
        }
```

In other words, if the target value is too small for the param buffer size, then it's an error

Reply:
What happens if you get an int when the value passed in is a double?
What happens if you get a long when the value passed in is a size_t or unsigned long?
These are cases where the size is the same, but the bit representation differs.

Reply:
For doubles, we have `OSSL_PARAM_REAL`, so different data types.
For signed vs unsigned integers, we have `OSSL_PARAM_INTEGER` vs `OSSL_PARAM_UNSIGNED_INTEGER`

Reply:
Integer truncation don’t happen, only widening. Zero or sign extending as per C rules.

I’ve been more flexible with doubles and integers. I’m comfortable with these occurring though.